cdc/split.cc

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mutation.hh"
+#include "schema.hh"
+
+#include "split.hh"
+
+namespace cdc {
+
+bool should_split(const mutation& base_mutation, const schema& base_schema) {
+    auto& p = base_mutation.partition();
+
+    api::timestamp_type found_ts = api::missing_timestamp;
+    std::optional<gc_clock::duration> found_ttl; // 0 = "no ttl"
+
+    auto check_or_set = [&] (api::timestamp_type ts, gc_clock::duration ttl) {
+        if (found_ts != api::missing_timestamp && found_ts != ts) {
+            return true;
+        }
+        found_ts = ts;
+
+        if (found_ttl && *found_ttl != ttl) {
+            return true;
+        }
+        found_ttl = ttl;
+
+        return false;
+    };
+
+    bool had_static_row = false;
+
+    bool should_split = false;
+    p.static_row().get().for_each_cell([&] (column_id id, const atomic_cell_or_collection& cell) {
+        had_static_row = true;
+
+        auto& cdef = base_schema.column_at(column_kind::static_column, id);
+        if (cdef.is_atomic()) {
+            auto view = cell.as_atomic_cell(cdef);
+            if (check_or_set(view.timestamp(), view.is_live_and_has_ttl() ? view.ttl() : gc_clock::duration(0))) {
+                should_split = true;
```

Comment:
Better to use `mutation_partition_visitor` because when we extend the mutation model this code will become silently broken.

Reply:
@tgrabiec won't the type system and/or tests catch the breakage?

Also, if this code would become broken, won't `mutation_partition_visitor` be broken too?

By the way, `transform` in cdc/log.cc also has this problem because it walks around the mutation manually. But I plan to refactor it (we didn't in this PR so to not conflict with changes made concurrently e.g. by @elcallio), and then I'll see if `mutation_partition_visitor` is up to the job. Thanks for the heads-up.

Reply:
The type system won't catch the breakage if it's caused by adding new elements because you're currently explicitly accessing the elements you expect by calling appropriate getters.

`mutation_partition_visitor` is part of the mutation model, so it can be trusted by CDC. The visitor has higher chance of being up to date, since it's the foundation of many algorithms which consume the model.

Reply:
@tgrabiec is right. We're planning some serious refactoring after 4.0 is cut. We definitely should use mutation_partition_visitor then.